trinity/utils/datastructures.py

```diff
@@ -286,3 +297,315 @@ def complete(self, batch_id: int, completed: Tuple[TTask, ...]) -> None:
+
+        # this await would hang, waiting on the receipt from header1:
+        # await block_import_tasks.ready_tasks()
+
+        block_import_tasks.finish_prereq(BlockDownloads.receipts, (header1, ))
+
+        # now we have all the necessary info to import blocks 1 and 2
+        headers_ready_to_import = await block_import_tasks.ready_tasks()
+
+        # these will always return in sequential order:
+        assert headers_ready_to_import == (header1, header2)
+
+    In a real implementation, you would have a loop waiting on
+    :meth:`ready_tasks` and import them, rather than interleaving them like
+    the above example.
+
+    Note that this class does *not* track when the main tasks are
+    complete. It is assumed that the caller will complete the tasks in the
+    order they are returned by ready_tasks().
+
+    The memory needs of this class would naively be unbounded. Any newly
+    registered task might depend on any other task in history. To prevent
+    unbounded memory usage, old tasks are pruned after a configurable depth.
+
+    Vocab:
+
+    - prerequisites: all these must be completed for a task to be ready
+        (a necessary but not sufficient condition)
+    - ready: a task is ready after all its prereqs are completed, and the task it depends on is
+        also ready
+    """
+    # methods to extract the id and dependency IDs out of a task
+    _id_of: StaticMethod[Callable[[TTask], TTaskID]]
+    _dependency_of: StaticMethod[Callable[[TTask], TTaskID]]
+
+    # by default, how long should the integrator wait before pruning?
+    _default_max_depth = 10000  # not sure how to pick a good default here
+
+    _prereq_tracker: Type[BaseTaskPrerequisites[TTask, TSubtask]]
+
+    def __init__(
+            self,
+            prerequisites: Type[TSubtask],
+            id_extractor: Callable[[TTask], TTaskID],
+            dependency_extractor: Callable[[TTask], TTaskID],
+            max_depth: int = None) -> None:
+
+        self._prereq_tracker = BaseTaskPrerequisites.from_enum(prerequisites)
+        self._id_of = id_extractor
+        self._dependency_of = dependency_extractor
+        self._oldest_depth = 0
+
+        # how long to wait before pruning
+        if max_depth is None:
+            self._max_depth = self._default_max_depth
+        else:
+            self._max_depth = min([self._default_max_depth, max_depth])
+
+        # all of the tasks that have been completed, and not pruned
+        self._tasks: Dict[TTaskID, BaseTaskPrerequisites[TTask, TSubtask]] = {}
+
+        # In self._dependencies, when the key becomes ready, the task ids in the
+        # value set *might* also become ready
+        # (they only become ready if their prerequisites are complete)
+        self._dependencies: Dict[TTaskID, Set[TTaskID]] = defaultdict(set)
+
+        # task ids are in this set if they either:
+        # - are missing a prerequisite OR
```

Comment:
missing is confusing for me here, would saying "all of their prerequisites are not completed" be accurate?